Fix for issue 5752: pwsh cannot start in a directory with a name that includes a wildcard

JamesWTruher opened the pull request:
## PR Summary

Fix #5752.

If you attempt to start pwsh in a directory like `/tmp/[T]est` the shell will default to $PSHOME

The issue was that we attempted to glob against the string returned by `Directory.GetCurrentDirectory()` which we should not do.

Technically, this is a breaking change - anyone that relies on starting in $PSHOME under these circumstances may be surprised by this behavior.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed - Issue link: https://github.com/PowerShell/PowerShell/issues/5752
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively 
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

JamesWTruher committed:
Support pwsh starting in directories which may have wildcard characters in the name.

There's really no reason why you should glob a path when you're attempting to go to Directory.GetCurrentDirectory()

JamesWTruher committed:
Update ConsoleHost.Tests.ps1

JamesWTruher commented:
it looks like the test that failed is unrelated to my change. In reviewing the threadjob test failure, there looks to be a timing issue. I'll follow up with @PaulHigin.

iSazonov commented on test/powershell/Host/ConsoleHost.Tests.ps1:
```diff
@@ -683,3 +683,24 @@ Describe "Pwsh exe resources tests" -Tag CI {
+    BeforeAll {
+        $powershell = Join-Path -Path $PsHome -ChildPath "pwsh"
+        $dirnames = "[T]est","[Test","T][est","Test"
+        foreach ( $d in $dirnames ) { new-item -type Directory -path "${TESTDRIVE}/$d" }
```
Please suppress output `new-item -type Directory -path "${TESTDRIVE}/$d" > $null`.
Please use common formatting style:
```powershell
 foreach ( $d in $dirnames ) {
    new-item -type Directory -path "${TESTDRIVE}/$d" > $null
}
```

JamesWTruher replied:
sure, although this is slower than:
`$null = New-Item -type Directory -path "${TESTDRIVE}/$d"`

iSazonov commented on test/powershell/Host/ConsoleHost.Tests.ps1:
```diff
@@ -683,3 +683,24 @@ Describe "Pwsh exe resources tests" -Tag CI {
+        param ( $dirname )
+        try {
+            Push-Location -LiteralPath "${TESTDRIVE}/${dirname}"
+            $result = & $powershell -c '(Get-Item .).Name'
```
Should we use cmdlet? Maybe better .Net method to get current directory or -LiteralPath?

JamesWTruher replied:
it really doesn't matter, does it?

adityapatwardhan replied:
I agree, maybe use 
```PowerShell
$result = & $powershell -c '([Environment]::CurrentDirectory)'
```

JamesWTruher replied:
also, I want to check only the name,  ason some systems, Environment.CurrentDirectory will include a symlink which is a problem because PS does not, I'm I need to call split-path, I can use this form just as well.

iSazonov replied:
My misunderstanding is that we use `Directory.GetCurrentDirectory()` in the code and `Set-Location` and `Get-Item` in the test - do we work with different current directories (first in .Net, second in FileSystem provider)?

KillyMXI commented on src/System.Management.Automation/engine/InitialSessionState.cs:
```diff
@@ -3229,6 +3229,7 @@ internal static void SetSessionStateDrive(ExecutionContext context, bool setLoca
 
                         try
                         {
+                            providerContext.SuppressWildcardExpansion = true;
```
(style) Maybe it's just my OCD, by why configuring providerContext inside try block rather than right after initialization?

JamesWTruher replied:
I wanted the two operations to be visually close, putting the assignment in the block did that for me

adityapatwardhan commented on test/powershell/Host/ConsoleHost.Tests.ps1:
```diff
@@ -683,3 +683,24 @@ Describe "Pwsh exe resources tests" -Tag CI {
+        $powershell = Join-Path -Path $PsHome -ChildPath "pwsh"
+        $dirnames = "[T]est","[Test","T][est","Test"
+        foreach ( $d in $dirnames ) { new-item -type Directory -path "${TESTDRIVE}/$d" }
+        $testcases = foreach ( $d in $dirnames ) { @{ Dirname = $d } }
```
We iterate over `$dirname` twice, once for `new-item` and once for generating test cases. Should be do it in a single loop?

JamesWTruher replied:
sure

BrucePay reviewed (approved):
LGTM

JamesWTruher committed:
Address a timing issue in ThreadJobs tests.

Sometimes the job object will return with state Running before it is actually running

PaulHigin reviewed (approved):
Looking at these changes I realize that we really only need wait-forrunning function for the Job2 tests because they test job state and/or use job APIs with the assumption of the job running.  And also for the couple of other tests that wait for job state to go to running (with throttle limit).  Otherwise Wait-Job is used which will wait for an end state.  I don't think it hurts anything but it is not needed.

JamesWTruher commented:
@adityapatwardhan using the API approach will fail in some circumstances where `${TESTDRIVE}/...` in PowerShell is `/tmp/<guid>/...` but the API may return `/private/tmp/<guid>/...` as it follows the symlink. Testing the name of the directory is the easiest way to avoid the necessity of recognizing/constructing the "real" path. What I've done proves the behavior well enough.

adityapatwardhan commented:
@JamesWTruher Is the thread jobs test fix intended to be in this PR? If so could you update the title and description.

anmenaga commented:
@JamesWTruher it looks like job test fix was moved to a separate PR #7270 (with some additions), so they can be excluded from this one.

JamesWTruher committed:
remove changes in threadjob tests

ZeppLu commented:
Sorry to bother, but when will this fix be in Windows? I'm using PowerShell 5.1.17763.134 on Windows 1809, and issue 5752 still exists.

vexx32 commented:
@ZeppLu the PS team have mentioned in previous issues and PRs that no more fixes are being backported to Windows PowerShell.

KillyMXI commented:
This fix was included into the version v6.1.0-preview.4 - 2018-07-19.
The question is, when the fresh enough version of PowerShell will be shipped with Windows.

ZeppLu commented:
@vexx32 Installing PowerShell Core does not seem a good solution for me, I'll try out other workarounds. Anyway, thanks for your explanation and advice.
